contracts/managers/allocators/SocialAllocator.sol

```diff
@@ -0,0 +1,401 @@
+        uint256 minDecimals = Math.min(baseAssetDecimals, quoteAssetDecimals);
+
+        // Decimal difference for asset a1 is 10 ** (a1Decimals - min(a1Decimals, a2Decimals))
+        baseAssetDecimalDifference = CommonMath.safePower(10, uint256(baseAssetDecimals).sub(minDecimals));
+        quoteAssetDecimalDifference = CommonMath.safePower(10, uint256(quoteAssetDecimals).sub(minDecimals));
+
+        // NaturalUnit is equal to max(a1DecimalDifference, a2DecimalDifference) * pricePrecision
+        collateralNaturalUnit = Math.max(baseAssetDecimalDifference, quoteAssetDecimalDifference)
+            .mul(pricePrecision);
+
+        // Next set components will always be in order of base asset first
+        multiAssetNextSetComponents = [address(baseAsset), address(quoteAsset)];
+        baseAssetNextSetComponents = [address(baseAsset)];
+        quoteAssetNextSetComponents = [address(quoteAsset)];
+    }
+
+    /* ============ External ============ */
+
+    /*
+     * Determine the next allocation to rebalance into.
+     *
+     * @param  _targetBaseAssetAllocation       Target allocation of the base asset
+     * @param  _allocationPrecision             Precision of allocation percentage
+     * @return ISetToken                        The address of the proposed nextSet
+     */
+    function determineNewAllocation(
+        uint256 _targetBaseAssetAllocation,
+        uint256 _allocationPrecision
+    )
+        external
+        returns (ISetToken)
+    {
+        // Determine nextSet units and components
+        (
+            uint256[] memory nextSetUnits,
+            address[] memory nextSetComponents
+        )= calculateNextSetParameters(
+            _targetBaseAssetAllocation,
+            _allocationPrecision
+        );
+
+        // Create new collateral set with passed components, units, and collateralNaturalUnit
+        address nextSetAddress = core.createSet(
+            setTokenFactory,
+            nextSetComponents,
+            nextSetUnits,
+            collateralNaturalUnit,
+            "",
+            "",
+            ""
+        );
+
+        return ISetToken(nextSetAddress);
+    }
+
+    /*
+     * Calculate value of passed collateral set.
+     *
+     * @param  _collateralSet         of current set collateralizing RebalancingSetToken
+     * @return uint256               USD value of passed Set
+     */
+    function calculateCollateralSetValue(
+        ISetToken _collateralSet
+    )
+        external
+        view
+        returns (uint256)
+    {
+        return calculateCollateralSetValueInternal(
+            address(_collateralSet)
+        );
+    }
+
+    /* ============ Internal ============ */
+
+    /*
+     * Calculate value of passed collateral set.
+     *
+     * @param  _currentCollateralSet        Instance of current set collateralizing RebalancingSetToken
+     * @return uint256                      USD value of passed Set
+     */
+    function calculateCollateralSetValueInternal(
```

Comment:
In `set-protocol-contracts`, there's a `SetUSDValuation` file. Perhaps we can standardize the code for calculating USD price of a Set, so we don't need to rewrite/reimplement across codebases.

The other use that seems to clean things up is also using `oracleWhitelist` vs. individual oracleAddresses passed in the constructor and held in state